src/dev/flang/ast/Expr.java

```diff
@@ -223,41 +223,50 @@ AbstractType typeForUnion()
    * @return the union of exprs result type, defaulting to Types.resolved.t_void if
    * no expression can be inferred yet.
    */
-  public static AbstractType union(List<Expr> exprs)
+  public static AbstractType union(List<Expr> exprs, boolean assumeVoidIfUnknown)
   {
     AbstractType t = Types.resolved.t_void;
 
     // First pass:
     // Union of the types of the expressions
     // that are sure about their types.
+    var foundType = false;
     for (var e : exprs)
       {
         var et = e.typeForUnion();
         if (et != null)
           {
+            foundType = true;
             t = t.union(et);
           }
       }
 
-    // Propagate the found type to all expression
-    for (var e : exprs)
+    if (foundType)
       {
-        e.propagateExpectedType(t);
+        // Propagate the found type to all expression
+        for (var e : exprs)
+          {
+            e.propagateExpectedType(t);
+          }
       }
 
     // Second pass:
     // Union of the types of the expressions
     AbstractType result = Types.resolved.t_void;
+    foundType = false;
     for (var e : exprs)
       {
         var et = e.typeForInferencing();
         if (et != null)
           {
+            foundType = true;
             result = result.union(et);
           }
       }
 
-    return result;
+    return foundType || assumeVoidIfUnknown
+      ? result
+      : null;
```

Comment:
If you cannot remove `assumeVoidIfUnknown`: I think the code would be clearer if you would introduce a second boolean variable here, e.g., 

```
    var gotResult = assumeVoidIfUnknown;
    for (var e : exprs)
      {
        var et = e.typeForInferencing();
        if (et != null)
          {
            gotResult = true;
            result = result.union(et);
          }
      }

    return result;
    return gotResult
      ? result
      : null;
```